### ISSUE TITLE
2 Factor Authentication code does not match

### ISSUE THREAD

**Issue by dhuyvetter** [jasonmunro/cypht#148]
When I enable 2 Factor Authentication and try to log in again, I keep getting the "2 factor authentication code does not match" error. The only way to log in is with one of the backup codes. I have deleted the Cypht entry in Google Authenticator and add it again, but still doesn't work.

**Comment by jasonmunro** [comment]
Hello! The likely issue here is time skew between your 2FA client and your server. When my server time is not accurate enough, 2FA stops working. You can use the ntp daemon to keep your server time accurate, or to sync it up one time the following command:

`sudo ntpdate pool.ntp.org`

If that doesn't help let me know and we can dig into it more. Thanks for the feedback!

**Comment by manasb** [comment]
Yeah, my Google Authenticator login keys often do not work when the time on my Android is inaccurate.

**Comment by jasonmunro** [comment]
@manasb exactly. The times need to be in sync between the client and the server, since it's a part of what is used to calculate the 6 digit code.

**Comment by dhuyvetter** [comment]
Thanks @jasonmunro I re-synced the time on my server and now it is working!

**Comment by dhuyvetter** [comment]
Time kept going off, adding `ntpdate pool.ntp.org` to `crontab` didn't help, I still needed to do it manually every couple of days. So installed `sudo apt install ntp`: https://help.ubuntu.com/lts/serverguide/NTP.html

**Comment by conspacer** [comment]
@jasonmunro maybe should add a small note on this in the 2fa.ini (like: "make sure your server is synced via [NTP](http://www.ntp.org/) before enabling 2FA" or something to that effect).
Spent 15 minutes trying to figure out why 2FA was not working correctly until I found this thread :)

**Comment by jasonmunro** [comment]
@conspacer Great idea, I just added the following:

; In order for 2fa to work, your server MUST have an accurate date and time,
; otherwise the codes won't match up. NTP is the standard way to keep a server's
; time synced: http://www.ntp.org/

**Comment by jasonmunro** [comment]
@conspacer I just did some testing and found that 2fa was not working for me in the current master or release branch. The passcodes were being accepted, but then when I was logged in, I lost all my previously saved settings! Not sure if ran into this, it was a bug I introduced here: https://github.com/jasonmunro/cypht/commit/c55ef123f4d97483c8126e01eded5c6108347f87 Anyway, it's fixed now in both the release and master branch.
